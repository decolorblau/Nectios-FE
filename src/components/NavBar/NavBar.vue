<template>
  <v-app-bar density="comfortable" flat app>
    <v-app-bar-title class="text-uppercase grey--text text-h4" to="/">
      <router-link to="/" class="title display-1">
        <span class="font-weight-light">BLUE</span>
        <span class="world blue">WORLD</span>
      </router-link>
    </v-app-bar-title>
    <v-spacer></v-spacer>
    <v-list class="hidden-sm-and-down">
      <v-btn flat @click="redirectToHome()" color="none" class="mr-4 text-uppercase">
        <span class="text-bar">Home</span>
      </v-btn>

      <v-btn
        to="/user"
        @click="redirectToUser()"
        flat
        :ripple="false"
        color="none"
        class="mr-4 text-uppercase"
      >
        <span class="text-bar">User</span>
      </v-btn>

      <v-btn flat @click="redirectToNewProduct()" color="none" class="mr-4 text-uppercase">
        <p class="text-bar"><span class="blue">+ADD</span>PRODUCT</p>
      </v-btn>
      <v-divider class="mx-4" vertical inset />

      <v-btn class="mr-4 ml-5" to="/login" color="none" flat @click="logout()">
        <span class="text-bar">Logout</span>
      </v-btn>
    </v-list>
    <v-banner-content class="hidden-md-and-up">
      <v-btn @click="showDrawer()" color="none" flat text>
        <em class="fas fa-bars"></em>
      </v-btn>
    </v-banner-content>
  </v-app-bar>
  <v-navigation-drawer
    v-model="drawer"
    position="right"
    width="1000"
    elevation="0"
    color="white"
    border="none"
    class="hidden-md-and-up"
  >
    <v-list nav dense align="center">
      <div active-class="deep-purple--text text--accent-4">
        <v-list-item
          class="pl-7 text-bar"
          title="Home"
          @click="redirectToHome(), showDrawer()"
        ></v-list-item>
        <v-list-item
          class="pl-7 text-bar"
          title="User"
          @click="redirectToUser(), showDrawer()"
        ></v-list-item>
        <v-list-item
          class="pl-7 text-bar"
          title="+ ADD PRODUCT"
          @click="redirectToNewProduct(), showDrawer()"
        ></v-list-item>
        <v-spacer></v-spacer>
        <v-list-item
          class="pl-7 text-bar"
          title="Logout"
          @click="logout(), showDrawer()"
        ></v-list-item>
      </div>
    </v-list>
  </v-navigation-drawer>
</template>

<script>
import { defineComponent } from "vue";
import { mapActions, mapGetters } from "vuex";

export default defineComponent({
  name: "NavBar",
  data() {
    return {
      drawer: false,
    };
  },

  methods: {
    ...mapActions(["loginUser", "logoutUser"]),
    ...mapGetters(["redirectToHome", "redirectToUser", "redirectToLogin", "redirectToNewProduct"]),

    logout() {
      this.logoutUser();
      this.redirectToLogin();
    },
    showDrawer() {
      this.drawer = !this.drawer;
    },
  },
});
</script>

<style lang="scss" scoped>
.title {
  color: inherit;
  text-decoration: none;

  :visited {
    color: inherit;
  }
}
.text-bar {
  font-weight: 500;
}
.blue {
  color: #207eeb;
}
.world {
  font-weight: 600;
}
.vertical-line {
  width: 1px;
  height: 30px;
  background-color: rgb(80, 80, 80);
  margin: 0;
  padding: 0;
}
</style>
