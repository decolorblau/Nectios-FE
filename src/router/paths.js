const paths = {
  home: "/",
  detail: "/products/:id",
  login: "/login",
  newProduct: "/new",
  user: "/user",
  notFound: "/:pathMatch(.*)*",
};

export default paths;
