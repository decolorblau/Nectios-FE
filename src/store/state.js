const state = {
  products: [],
  comments: [],
  user: {
    isAuthenticated: false,
    clientKey: "",
    user: {
      email: "",
      name: "",
      profilePhoto: "",
      role: "",
      surname: "",
    },
  },
  currentProduct: {
    product: {
      id: "",
      client_id: "",
      description: "",
      key: "",
      name: "",
      price: "",
      product_photo: "",
    },
    comments: [],
  },
};

export default state;
