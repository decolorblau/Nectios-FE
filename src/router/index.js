import { createRouter, createWebHistory } from "vue-router";
import Home from "../views/Home.vue";
import Login from "../views/LoginView.vue";
import NotFound from "../views/NotFoundView.vue";
import paths from "./paths";
import User from "../views/UserView.vue";
import Detail from "../views/DetailView.vue";

// eslint-disable-next-line object-curly-newline
const { home, notFound, login, user, detail } = paths;

const routes = [
  {
    path: home,
    name: "Home",
    component: Home,
  },
  {
    path: login,
    name: "Login",
    component: Login,
  },
  {
    path: user,
    name: "User",
    component: User,
  },
  {
    path: detail,
    name: "Detail",
    component: Detail,
  },
  {
    path: notFound,
    name: "NotFound",
    component: NotFound,
  },
];

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

export default router;
