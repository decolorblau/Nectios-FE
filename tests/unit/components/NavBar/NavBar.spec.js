import NavBar from "@/components/NavBar/NavBar.vue";
import { shallowMount, mount } from "@vue/test-utils";
import { createStore } from "vuex";
import router from "@/router";
import state from "../../mockState";

describe("Given a NavBar component", () => {
  describe("When is rendered", () => {
    test("Then it matches de snapshot", () => {
      const wrapper = shallowMount(NavBar);
      expect(wrapper.html()).toMatchSnapshot();
    });
    test("Then it should render BLUEWORLD input", () => {
      const wrapper = shallowMount(NavBar);
      const [blue, world] = wrapper.findAll("span");
      expect(blue.text()).toBe("BLUE");
      expect(world.text()).toBe("WORLD");
    });
    describe("When we are in desktop Design", () => {
      test("Then it should render Home button", () => {
        const wrapper = shallowMount(NavBar);
        const [homeButton] = wrapper.findAll("v-btn");
        expect(homeButton.text()).toBe("Home");
      });
      describe("When the Home is clickable", () => {
        test("Then it should invoke redirectToHome", async () => {
          const store = createStore({
            state() {
              return state;
            },
            actions: {
              loginUser: jest.fn(),
              logoutUser: jest.fn(),
            },
          });

          const wrapper = mount(NavBar, {
            global: {
              plugins: [router, store],
            },
          });

          const redirectToHome = jest.fn();
          redirectToHome();
          const [homeButton] = wrapper.findAll("v-btn");
          await homeButton.trigger("click");
          expect(redirectToHome).toHaveBeenCalled();
        });
      });
      test("Then it should render User button", () => {
        const wrapper = shallowMount(NavBar);
        const [, userButton] = wrapper.findAll("v-btn");
        expect(userButton.text()).toBe("User");
      });
      describe("When the User is clickable", () => {
        test("Then it should invoke redirectToUser", async () => {
          const store = createStore({
            state() {
              return state;
            },
            actions: {
              loginUser: jest.fn(),
              logoutUser: jest.fn(),
            },
          });

          const wrapper = mount(NavBar, {
            global: {
              plugins: [router, store],
            },
          });

          const redirectToUser = jest.fn();
          redirectToUser();
          const [, userButton] = wrapper.findAll("v-btn");
          await userButton.trigger("click");
          expect(redirectToUser).toHaveBeenCalled();
        });
      });
      test("Then it should render +ADDPRODUCT button", () => {
        const wrapper = shallowMount(NavBar);
        const [, , addButton] = wrapper.findAll("v-btn");
        expect(addButton.text()).toBe("+ADDPRODUCT");
      });
      describe("When the +ADDPRODUCT is clickable", () => {
        test("Then it should invoke redirectToUser", async () => {
          const store = createStore({
            state() {
              return state;
            },
            actions: {
              loginUser: jest.fn(),
              logoutUser: jest.fn(),
            },
          });

          const wrapper = mount(NavBar, {
            global: {
              plugins: [router, store],
            },
          });

          const redirectToNewProduct = jest.fn();
          redirectToNewProduct();
          const [, , addButton] = wrapper.findAll("v-btn");
          await addButton.trigger("click");
          expect(redirectToNewProduct).toHaveBeenCalled();
        });
      });
      test("Then it should render logout button", () => {
        const wrapper = shallowMount(NavBar);
        const [, , , logout] = wrapper.findAll("v-btn");
        expect(logout.text()).toBe("Logout");
      });
      describe("When the logout is clickable", () => {
        test("Then it should invoke redirectToUser", async () => {
          const store = createStore({
            state() {
              return state;
            },
            actions: {
              loginUser: jest.fn(),
              logoutUser: jest.fn(),
            },
            getters: {
              redirectToLogin: jest.fn(),
            },
          });

          const wrapper = mount(NavBar, {
            global: {
              plugins: [router, store],
            },
          });

          const logout = jest.fn();
          logout();
          const [, , , logoutButton] = wrapper.findAll("v-btn");
          await logoutButton.trigger("click");
          expect(logout).toHaveBeenCalled();
        });
      });
    });
    describe("When we are in mobile Design", () => {
      test("Then it should render Home button", () => {
        const wrapper = shallowMount(NavBar);
        const [homeButton] = wrapper.findAll("v-list-item");
        expect(homeButton.attributes("title")).toBe("Home");
      });
      describe("When the Home is clickable", () => {
        test("Then it should invoke redirectToHome", async () => {
          const store = createStore({
            state() {
              return state;
            },
            actions: {
              loginUser: jest.fn(),
              logoutUser: jest.fn(),
            },
          });

          const wrapper = mount(NavBar, {
            global: {
              plugins: [router, store],
            },
          });

          const redirectToHome = jest.fn();
          redirectToHome();
          const showDrawer = jest.fn();
          showDrawer();
          const [homeButton] = wrapper.findAll("v-list-item");
          await homeButton.trigger("click");
          expect(redirectToHome).toHaveBeenCalled();
          expect(showDrawer).toHaveBeenCalled();
        });
      });
      test("Then it should render User button", () => {
        const wrapper = shallowMount(NavBar);
        const [, userButton] = wrapper.findAll("v-list-item");
        expect(userButton.attributes("title")).toBe("User");
      });
      describe("When the User is clickable", () => {
        test("Then it should invoke redirectToUser", async () => {
          const store = createStore({
            state() {
              return state;
            },
            actions: {
              loginUser: jest.fn(),
              logoutUser: jest.fn(),
            },
          });

          const wrapper = mount(NavBar, {
            global: {
              plugins: [router, store],
            },
          });

          const redirectToUser = jest.fn();
          redirectToUser();
          const showDrawer = jest.fn();
          showDrawer();
          const [, userButton] = wrapper.findAll("v-list-item");
          await userButton.trigger("click");
          expect(redirectToUser).toHaveBeenCalled();
          expect(showDrawer).toHaveBeenCalled();
        });
      });
      test("Then it should render +ADD PRODUCT button", () => {
        const wrapper = shallowMount(NavBar);
        const [, , addButton] = wrapper.findAll("v-list-item");
        expect(addButton.attributes("title")).toBe("+ ADD PRODUCT");
      });
      describe("When the +ADD PRODUCT is clickable", () => {
        test("Then it should invoke redirectToUser", async () => {
          const store = createStore({
            state() {
              return state;
            },
            actions: {
              loginUser: jest.fn(),
              logoutUser: jest.fn(),
            },
          });

          const wrapper = mount(NavBar, {
            global: {
              plugins: [router, store],
            },
          });

          const redirectToNewProduct = jest.fn();
          redirectToNewProduct();
          const showDrawer = jest.fn();
          showDrawer();
          const [, , addButton] = wrapper.findAll("v-list-item");
          await addButton.trigger("click");
          expect(redirectToNewProduct).toHaveBeenCalled();
          expect(showDrawer).toHaveBeenCalled();
        });
      });
      test("Then it should render logout button", () => {
        const wrapper = shallowMount(NavBar);
        const [, , , logout] = wrapper.findAll("v-list-item");
        expect(logout.attributes("title")).toBe("Logout");
      });
      describe("When the logout is clickable", () => {
        test("Then it should invoke redirectToUser", async () => {
          const store = createStore({
            state() {
              return state;
            },
            actions: {
              loginUser: jest.fn(),
              logoutUser: jest.fn(),
            },
            getters: {
              redirectToLogin: jest.fn(),
            },
          });

          const wrapper = mount(NavBar, {
            global: {
              plugins: [router, store],
            },
          });

          const logout = jest.fn();
          logout();
          const showDrawer = jest.fn();
          showDrawer();
          const [, , , logoutButton] = wrapper.findAll("v-list-item");
          await logoutButton.trigger("click");
          expect(logout).toHaveBeenCalled();
          expect(showDrawer).toHaveBeenCalled();
        });
      });
    });
  });
});
