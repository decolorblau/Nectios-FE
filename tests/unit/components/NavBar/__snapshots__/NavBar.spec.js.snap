// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a NavBar component When is rendered Then it matches de snapshot 1`] = `
<v-app-bar density="comfortable" flat="" app="">
  <v-app-bar-title class="text-uppercase grey--text text-h4" to="/">
    <router-link to="/" class="title display-1"><span class="font-weight-light">BLUE</span><span class="world blue">WORLD</span></router-link>
  </v-app-bar-title>
  <v-spacer></v-spacer>
  <v-list class="hidden-sm-and-down">
    <v-btn flat="" color="none" class="mr-4 text-uppercase"><span class="text-bar">Home</span></v-btn>
    <v-btn to="/user" flat="" ripple="false" color="none" class="mr-4 text-uppercase"><span class="text-bar">User</span></v-btn>
    <v-btn flat="" color="none" class="mr-4 text-uppercase">
      <p class="text-bar"><span class="blue">+ADD</span>PRODUCT</p>
    </v-btn>
    <v-divider class="mx-4" vertical="" inset=""></v-divider>
    <v-btn class="mr-4 ml-5" to="/login" color="none" flat=""><span class="text-bar">Logout</span></v-btn>
  </v-list>
  <v-banner-content class="hidden-md-and-up">
    <v-btn color="none" flat="" text=""><em class="fas fa-bars"></em></v-btn>
  </v-banner-content>
</v-app-bar>
<v-navigation-drawer modelvalue="false" position="right" width="1000" elevation="0" color="white" border="none" class="hidden-md-and-up">
  <v-list nav="" dense="" align="center">
    <div active-class="deep-purple--text text--accent-4">
      <v-list-item class="pl-7 text-bar" title="Home"></v-list-item>
      <v-list-item class="pl-7 text-bar" title="User"></v-list-item>
      <v-list-item class="pl-7 text-bar" title="+ ADD PRODUCT"></v-list-item>
      <v-spacer></v-spacer>
      <v-list-item class="pl-7 text-bar" title="Logout"></v-list-item>
    </div>
  </v-list>
</v-navigation-drawer>
`;
